programs/bpf_loader/src/serialization.rs

```diff
@@ -427,33 +460,31 @@ fn serialize_parameters_aligned(
                 s.write::<u64>((borrowed_account.get_rent_epoch()).to_le());
             }
             SerializeAccount::Duplicate(position) => {
+                account_lengths.push(account_lengths.get(position as usize).unwrap().clone());
                 s.write::<u8>(position as u8);
                 s.write_all(&[0u8, 0, 0, 0, 0, 0, 0]);
             }
         };
     }
-    s.write::<u64>((instruction_context.get_instruction_data().len() as u64).to_le());
-    s.write_all(instruction_context.get_instruction_data());
-    s.write_all(
-        instruction_context
-            .try_borrow_last_program_account(transaction_context)?
-            .get_key()
-            .as_ref(),
-    );
-
-    Ok(s.finish())
+    s.write::<u64>((instruction_data.len() as u64).to_le());
+    s.write_all(instruction_data);
+    s.write_all(program_id.as_ref());
+
+    let (mem, regions) = s.finish();
+    Ok((mem, regions, account_lengths))
 }
 
-pub fn deserialize_parameters_aligned(
+pub fn deserialize_parameters_aligned<I: IntoIterator<Item = usize>>(
     transaction_context: &TransactionContext,
     instruction_context: &InstructionContext,
     copy_account_data: bool,
     buffer: &[u8],
-    account_lengths: &[usize],
+    account_lengths: I,
```

Comment:
nit: i prefer existential types for these cases for one less type parameter (slight readability improvement, imo):

```
    account_lengths: impl IntoIterator<Item = usize>,
```